src/main/java/com/bugsnag/Bugsnag.java

```diff
@@ -39,11 +59,45 @@ public Bugsnag(String apiKey, boolean sendUncaughtExceptions) {
         }
 
         config = new Configuration(apiKey);
+        sessionTracker = new SessionTracker(config);
 
         // Automatically send unhandled exceptions to Bugsnag using this Bugsnag
         if (sendUncaughtExceptions) {
             ExceptionHandler.enable(this);
         }
+        addSessionTrackingShutdownHook();
+        scheduleSessionFlushes();
+    }
+
+    private void scheduleSessionFlushes() {
+        sessionExecutorService.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                sessionTracker.flushSessions(new Date());
+            }
+        }, SESSION_TRACKING_PERIOD_SECS, SESSION_TRACKING_PERIOD_SECS, TimeUnit.SECONDS);
+    }
+
+    private void addSessionTrackingShutdownHook() {
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                sessionTracker.setShuttingDown(true);
```

Comment:
That's correct. Looking back at this implementation, it might make sense to adhere to the following logic:

- Only collect a session in `startSession` if `shuttingDown` is false
- When `shuttingDown` is set to true, attempt to deliver any remaining sessions

What are your thoughts?

Reply:
Updated to follow the logic laid out above.